### ISSUE TITLE
Supervisor Spring Cleaning

### ISSUE THREAD

**Issue by bodymindarts** [habitat-sh/habitat#2094]
Tl;Dr Some suggestions of general points of improvements hoping to trigger discussions and later concrete refactorings.

In this issue I would like to document a bunch of thoughts I am having around refactoring the code under `components/sup/src`.

As I would like to continue contributing by extending the functionality of the supervisor I also want to make an effort to pay down some technical debt to make working with the code easier and keep it more maintainable.

These thoughts are in no way aimed at or intended to lay any kind of blame or shame on any individual contributor or group. As every code base this one has grown over iterations to provide ever more and better functionality to the end user. A certain degradation of Internal Quality over time is unavoidable and some decisions made in haste just stick without being revisited. I have no doubt that every contributor made a best effort and that I would have done no better given the constraints that were faced at the time of each decision.

Also these suggestions are entirely my opinions on the largest areas that require improvements. I do not claim that they will lead to a 'perfect' design as that is not the goal. Nor that there aren't other or better ways to achieve the improvements.

These are however changes that I think will make working with the code more fun (at least for me). And as this is Open Source, having fun working on this is really my biggest motivator :-).

First I will make some general observations that hopefully will lead to discussions and agreement regarding how we want to handle these issues. Then I will make some concrete suggestions of things that I am willing to contribute.

Okay so general observations:

1) public / private and fields / methods

There is no agreement / style around when methods or fields should be public and or private. In actual fact for the most part all fields and methods are public. This is of serious concern to me because it inflates the public interface of objects which in turn encourages (though unconsciously) arbitrary coupling to object internals. This makes any kind of refactoring very difficult since it is impossible to know which methods / fields are really required by clients and which are just arbitrarily public. I advocate to take the [minimal interface](https://martinfowler.com/bliki/MinimalInterface.html), and Intention Revealing Interface approach in our designs.
There are also many examples of using rusts `Deref` trait that unnecessarily bloat the interfaces in a non behaviour driven way. This may be convenient but I think for long term maintainability we are better of considering only the minimal set of what the client needs to avoid confusion.

2) configuration / data ownership

There is a lot of duplication between `Service`, `ServiceSpec`, `PackageInstall`, `ServiceConfig` and `RuntimeConfig` in terms of what data they hold and reveal. I think we should refactor this to have clearer boarders in terms of concepts and data ownership and be more consistent as to what objects we deal with when it comes to retrieving information. IMO it would be good if there were conceptually 1 source of truth for each relevant piece of information regarding a service, or if we want to decouple some parts of the object graph then an explicit location of translation from 1 interface to the other.

3) Respecting of layers

There is some hint of a layered approach with the `butterfly::server` code being responsible for the network protocol and the `manager::service` code taking care of the domain logic where the  `Manager` exists as an application entrypoint and coordination layer.
There is even an adaptation of the underlying objects used by the butterfly code (the `Rumors`) to a more application specific view (the `Census`). I really like this approach but unfortunately the execution is lacking. This is most visible in the `Service` code that should represent the application / domain layer but has knowledge of internal details of both the `Census` as well as the `butterfly` concepts.

4) Serialization

There are many places in the code where we are serializing and deserializing or rendering various objects to various places for various reasons. This concern is spread throughout the code base and details are contained in almost every object.
Since these operations are so frequent I feel that when and where data is persisted should be visible as a first class concern in the code base and not just added to the objects via a bunch of methods.
At the moment it is particularly hard to reason about the lifecycle of the objects because it is not explicit what the triggers for these methods are and from where they are being called (again because they are all public).

**Comment by bodymindarts** [comment]
I am happy to take on refactorings that address some parts of the issues raised above.
First of all I have created https://github.com/habitat-sh/habitat/pull/2095 to make all unneeded public methods and fields private. This somewhat addresses 1) though it isn't intended to improve the design of the code. Just to make refactoring easier when trying to figure out the real public interface.

**Comment by bodymindarts** [comment]
Another step in the ongoing improvement effort https://github.com/habitat-sh/habitat/pull/2141

**Comment by bodymindarts** [comment]
Here we go... remove `butterfly::Service` from `Service`. This is mostly addressing the layers topic https://github.com/habitat-sh/habitat/pull/2171.

**Comment by bodymindarts** [comment]
I'm going to declare this as finished... its hardly spring anymore and most issues have been addressed or at least significantly improved... hurray.

**Comment by bodymindarts** [comment]
Thanks @reset for the help!
